File: src/main/java/prosia/basarnas/web/util/map/poly/TSRPattern.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package prosia.basarnas.web.util.map.poly;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import org.primefaces.model.map.LatLng;
import org.primefaces.model.map.Marker;
import org.primefaces.model.map.Polyline;
import prosia.basarnas.constant.CommonConstant;
import prosia.basarnas.web.util.map.CommonUtil;

/**
 *
 * @author Ismail
 */
public class TSRPattern implements Serializable {

    private CommonUtil commonUtil = new CommonUtil();
    private LatLng start;
    private double searchLeg;
    private double searchWidth;
    private double searchTrackSpacing;
    private double searchHeading;
    private LatLng pivot;
    private List<LatLng> latlngs;
    private String color;
    private int rotateState;

    @Getter
    private Polyline polyline = new Polyline();
    @Getter
    private List<Marker> markersWaypointToolTip;
    @Getter
    private String contentString;

    public TSRPattern(LatLng start, double searchTrackSpacing, double searchLeg,
            double searchHeading, LatLng pivot, List<LatLng> latlngs, String color) {
        this.color = color;
        this.start = start;
        this.searchTrackSpacing = searchTrackSpacing;
        this.searchLeg = searchLeg;
        this.searchHeading = searchHeading;
        if (pivot == null) {
            this.pivot = new LatLng((start.getLat()
                    + ((searchLeg / 2) * CommonConstant.ONE_NM_TO_DEG)),
                    (start.getLng() + ((searchTrackSpacing / 2) * CommonConstant.ONE_NM_TO_DEG)));
        } else {
            this.pivot = pivot;
        }
        if (latlngs == null || latlngs.isEmpty()) {
            init();
            firstHeadings();
            initWaypointToolTip();
        } else {
            this.latlngs = latlngs;
        }

        contentString = "<b class='title'><u>Track Search Return Search</u></b>"
                + "<br><b class='field'>Search Leg</b>      : <font class='value'>" + searchLeg + " NM</font>"
                + "<br><b class='field'>Heading </b> : <font class='value'>" + searchHeading + " °</font>"
                + "<br><b class='field'>Track Spacing </b> : <font class='value'>" + searchTrackSpacing + " NM</font>";
        //+   "<br><b class='field'>Length</b> : <font class='value'>" + this.getContentStringOfLength() + "</font>";

    }

    private void init() {
        List<LatLng> listLatLngs = new ArrayList<>();
        listLatLngs.add(start);
        LatLng point1 = new LatLng(start.getLat()
                + (searchLeg * CommonConstant.ONE_NM_TO_DEG), this.start.getLng());
        LatLng point2 = new LatLng(point1.getLat(), point1.getLng()
                + (searchTrackSpacing * CommonConstant.ONE_NM_TO_DEG));
        LatLng point3 = new LatLng(this.start.getLat(), point2.getLng());
        rotateState = 0;
        listLatLngs.add(point1);
        listLatLngs.add(point2);
        listLatLngs.add(point3);
        createAllPolyLine(listLatLngs);
    }

    public void firstHeadings() {
        LatLng firstHeadingPivot = new LatLng(start.getLat(), (start.getLng()
                + (searchTrackSpacing / 2 * CommonConstant.ONE_NM_TO_DEG)));
        List<LatLng> rotation = commonUtil.rotationLatLngs(polyline.getPaths(), firstHeadingPivot, searchHeading);
        polyline.setPaths(rotation);
        pivot = commonUtil.rotationLatLng(firstHeadingPivot, pivot, searchHeading);
    }

    public void headings() {
        List<LatLng> rotation = commonUtil.rotationLatLngs(polyline.getPaths(), pivot, searchHeading);
        polyline.setPaths(rotation);
    }
    
    public String inFocus() {
        return pivot.getLat() + "," + pivot.getLng();
    }

    public Integer inZoom() {
        return 10;
    }

    public void initWaypointToolTip() {
        markersWaypointToolTip = new ArrayList<>();
        for (int i = 0; i < this.polyline.getPaths().size() - 1; i++) {
            markersWaypointToolTip.add(new Marker(polyline.getPaths().get(i), "Waypoint " + (i + 1)));
        }
    }

    private void createAllPolyLine(List<LatLng> listLatLng) {
        polyline = new Polyline();
        for (LatLng gLatLng : listLatLng) {
            polyline.getPaths().add(new LatLng(gLatLng.getLat(),
                    gLatLng.getLng()));
        }
        polyline.setStrokeColor(color);
        polyline.setStrokeWeight(3);
        polyline.setStrokeOpacity(1);
    }
}

File: src/main/webapp/resources/js/map/mapv3/poly/TSRPattern.js
/* 
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */


//createTSN(1,new GLatLng(1.5200548106444896,103.568115234375),1,5,45,new GLatLng(1.412973212770802,103.79470825195312),null)
function TSRPattern(start, trackSpacing, searchLeg, heading, pivot, latlngs, color){
    this.color = color;
    this.start = start;
    this.trackSpacing = trackSpacing;
    this.searchLeg = searchLeg;
    this.heading = heading;
    if(!pivot){
        this.pivot = new GLatLng((start.lat() + ((searchLeg/2)*ONE_NM_TO_DEG)), (start.lng() + ((trackSpacing/2)*ONE_NM_TO_DEG)));
    }else{
        this.pivot = pivot;
    }
    if(latlngs){
        this.l = latlngs;
        this.draw();
    }else{
        this.init();
        this.firstHeadings();
        this.draw();
    }
}

TSRPattern.prototype = new SearchPattern();

TSRPattern.prototype.firstHeadings = function(){
    var firstHeadingPivot = new GLatLng(this.start.lat(), (this.start.lng() + (this.trackSpacing / 2 * ONE_NM_TO_DEG)));  
    rotationGLatLngs(this.l, firstHeadingPivot, this.heading);
    this.pivot = rotationGLatLng(firstHeadingPivot, this.pivot, this.heading);
}

TSRPattern.prototype.init = function(){
    this.l = new Array();
    this.l.push(this.start);
    var point1 = new GLatLng(this.start.lat() + (this.searchLeg * ONE_NM_TO_DEG), this.start.lng());
    var point2 = new GLatLng(point1.lat(), point1.lng() + (this.trackSpacing * ONE_NM_TO_DEG));
    var point3 = new GLatLng(this.start.lat(), point2.lng())
    this.rotateState = 0;
    this.l.push(point1);
    this.l.push(point2);
    this.l.push(point3);
}

TSRPattern.prototype.headings = function(){
    rotationGLatLngs(this.l, this.pivot, this.heading);
}


TSRPattern.prototype.inFocus = function(){
    map.setCenter(this.pivot, 10);
}

TSRPattern.prototype.move = function(from, to){
    this.moving(from, to);
    this.start = this.l[0];
    this.end = this.l[this.l.length];
    this.pivot = to;
    this.remove();
    this.draw();
}

TSRPattern.prototype.getHeading = function(){
    return this.heading;
}

TSRPattern.prototype.getTrackSpacing = function(){
    return this.trackSpacing;    
}

TSRPattern.prototype.getSearchLeg = function(){
    return this.searchLeg;
}

TSRPattern.prototype.getPivot = function(){
    return this.pivot;
}

TSRPattern.prototype.initWaypointToolTip = function(){
    var i, mouseOverCallback, mouseOutCallback;
    if(this.waypointsMarkers){
        for(i in this.waypointsMarkers){
            this.waypointsMarkers[i].setLatLng(this.l[i]);
        }
    }else{
        this.waypointsMarkers = [];
        var me = this;
        for(i = 0; i < this.l.length - 1; i++){
            var waypointMarker = CustomLabelMarker(TRANSPARANT_ICON, this.l[i], '');
            waypointMarker.setTitle("" +(i + 1));
            mouseOverCallback = function(latlng, marker){
                changeMarkerWithLabelContent(marker, stringHtmlOnMap("Waypoint " + marker.getTitle()));
                marker.label.setMap(map);
            };
            mouseOutCallback = function(latlng, marker){
                marker.label.setMap(null);
            };
            measurementPointIntegration(waypointMarker, mouseOverCallback, mouseOutCallback);
            map.addOverlay(waypointMarker);
            map.removeOverlay(waypointMarker.label);
            this.waypointsMarkers.push(waypointMarker);
        }
    }
}

function getTSRStartForGenerateValue(heading, trackSpacing, lkp){
    var orginalStart = new GLatLng(lkp.lat(), (lkp.lng() - (trackSpacing/2 * ONE_NM_TO_DEG)));
    return rotationLatLng(lkp, orginalStart, heading);
}

TSRPattern.prototype.getContentString = function(){
    return "<b class='title'><u>Track Search Return Search</u></b>" +
    "<br><b class='field'>Search Leg</b>      : <font class='value'>" + this.searchLeg +  " NM</font>" +
    "<br><b class='field'>Heading </b> : <font class='value'>" + this.heading + " °</font>" +
    "<br><b class='field'>Track Spacing </b> : <font class='value'>" + this.trackSpacing + " NM</font>" +
    "<br><b class='field'>Length</b> : <font class='value'>" + this.getContentStringOfLength() + "</font>";
}